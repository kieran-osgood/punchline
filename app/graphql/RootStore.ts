import { JokeModel } from "app/graphql"
import { RootStore as RootStoreTree } from "app/models"
import { destroy, getEnv, getRoot, IAnyStateTreeNode, Instance, types } from "mobx-state-tree"
import { RootStoreBase } from "./RootStore.base"

export interface RootStoreType extends Instance<typeof RootStore.Type> {}

export const RootStore = RootStoreBase.props({
  accessToken: types.maybeNull(types.string),
  deepLinkJokeId: types.maybe(types.string),
})
  .views((self) => ({
    get root(): RootStoreTree {
      return getRoot(self)
    },
  }))
  .actions((self) => ({
    removeChild(item: IAnyStateTreeNode) {
      destroy(item)
    },
    setAuthorizationHeader() {
      getEnv(self).gqlHttpClient.setHeaders({ Authorization: "bearer " + self.accessToken })
    },
    setDeepLinkJoke(deepLinkInitialJoke: string | undefined) {
      self.deepLinkJokeId = deepLinkInitialJoke
    },
  }))
  .actions((self) => ({
    setBearerToken(token: string) {
      self.accessToken = token
      self.setAuthorizationHeader()
    },
    fetchInitialJokes(deepLinkInitialJoke?: string) {
      const query = self.queryJokes(
        {
          blockedCategoryIds: self.root.settings.blockedCategoryIds,
          jokeLength: self.root.settings.jokeLengthMaxEnum,
          deepLinkedJokeId: deepLinkInitialJoke,
        },
        (j) =>
          j.nodes((n) =>
            n.id.body.title.negativeRating.positiveRating.categories((c) => c.id.image.name),
          ),
        { fetchPolicy: "no-cache" },
      )
      self.setDeepLinkJoke(deepLinkInitialJoke)

      return query
    },
  }))
  .views((self) => ({
    get nonViewedJokes() {
      return [...self.jokes.values()].filter((x) => !x.viewed)
    },
    get topOfDeckJoke() {
      if (this.nonViewedJokes.length <= 10) {
        self.fetchInitialJokes()
      }
      if (this.nonViewedJokes.length === 0) {
        return JokeModel.create({ id: "-1" })
      }
      if (self.deepLinkJokeId) {
        const id = self.deepLinkJokeId
        self.setDeepLinkJoke(undefined)
        const deepJoke = self.jokes.get(id)
        return deepJoke ?? this.nonViewedJokes[this.nonViewedJokes.length - 1]
      }

      return this.nonViewedJokes[this.nonViewedJokes.length - 1]
    },
    get bookmarkedJokes() {
      return [...self.userJokeHistories.values()].filter((x) => x.bookmarked)
    },
  }))

export const nodes = (val: any, ...rest: string[]) =>
  `nodes {
    ${val}
    ${rest.join(" ")}
  } 
  `
