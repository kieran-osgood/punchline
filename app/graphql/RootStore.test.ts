import { createHttpClient } from "mst-gql"
import { RootStore } from "./RootStore"

test("can be created without any user", () => {
  const store = RootStore.create(undefined, {
    gqlHttpClient: createHttpClient(""),
  })
  expect(store).toMatchSnapshot()
})

// test("can create request", async () => {
//   /** This functions provides a mock implementation for loading todos from the graphql endpoint. Also verifies the incoming arguments */
//   function mockLoadTodos(query, variables) {
//     expect(query).toMatchInlineSnapshot(`
//                       "query todos { todos {
//                               __typename
//                       id
//                       text
//                       complete
//                             } }"
//                 `)
//     expect(variables).toEqual(undefined)
//     return {
//       data: {
//         todos: [
//           {
//             id: "a",
//             __typename: "Todo",
//             complete: true,
//             text: "Initially loaded todo, now updated"
//           },
//           {
//             id: "b",
//             __typename: "Todo",
//             complete: false,
//             text: "Another todo"
//           }
//         ]
//       }
//     }
//   }

//   /** This functions provides a mock implementation for mutating a todo. Also verifies the incoming arguments */
//   function mockToggleTodos(query, variables) {
//     expect(query).toMatchInlineSnapshot(`
//       "mutation toggleTodo($id: ID!) { toggleTodo(id: $id) {
//               __typename
//       id
//       complete
//             } }"
//     `)
//     expect(variables).toEqual({ id: "b" })
//     return {
//       data: {
//         toggleTodo: { id: "b", __typename: "Todo", complete: true }
//       }
//     }
//   }

//   /** This test will make two following two requests */
//   const mockResponses = [mockLoadTodos, mockToggleTodos]
//   const mockClient = {
//     request(query, variables) {
//       return Promise.resolve(mockResponses.shift()!(query, variables)) // return and remove the first mocked response
//     }
//   }

//   /** Create a store with some initial state */
//   const store = RootStore.create(
//     {
//       todos: {
//         a: {
//           id: "a",
//           __typename: "Todo",
//           complete: false,
//           text: "Initially loaded todo"
//         }
//       },
//     },
//     {
//       gqlHttpClient: mockClient
//     }
//   )

//   // initially, 1 todo
//   expect(store.toJSON()).toMatchSnapshot()

//   // Then, the 2 todos mocked above should be loaded and merged
//   await store.queryTodos()
//   expect(store.todos.size).toBe(2)
//   expect(store.todos.get("a").text).toBe("Initially loaded todo, now updated")

//   const todoB = store.todos.get("b")
//   expect(todoB.text).toBe("Another todo")
//   expect(todoB.complete).toBe(false)

//   // Finally, also toggle a todo
//   await todoB.toggle()
//   expect(todoB.complete).toBe(true)

//   const store = RootStore.create(undefined, {
//     gqlHttpClient: createHttpClient(""),
//   })
// })
