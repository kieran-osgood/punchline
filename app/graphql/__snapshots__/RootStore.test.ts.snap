// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be created without any user 1`] = `
Object {
  "__queryCache": Object {},
  "accessToken": null,
  "categories": Object {},
  "jokes": Object {},
  "userJokeHistories": Object {},
  "users": Object {},
}
`;
